File: crates/ward_runtime/py/wardscript/schema.py
"""Runtime descriptions of Wardscript types, and what they're for: JSON schemas for
model answers, and decoding and encoding JSON values.

Generated modules describe their records and enums with `record()` and `enum()`, and
refer to types with the descriptors below (`List(String)`, `Adt(Ticket)`, ...).
"""

from __future__ import annotations

import enum as _enum
from dataclasses import dataclass
from typing import Any as _Any, Callable

from .errors import DecodeError
from .values import Some, some, unwrap


class Type:
    """A Wardscript type."""


@dataclass(frozen=True)
class Prim(Type):
    name: str

    def __repr__(self) -> str:
        return self.name


Int = Prim("Int")
Float = Prim("Float")
String = Prim("String")
Bool = Prim("Bool")
Unit = Prim("()")
#: Tool results, whose types aren't known yet.
Any = Prim("dynamic")


@dataclass(frozen=True)
class List(Type):
    elem: Type

    def __repr__(self) -> str:
        return f"List<{self.elem!r}>"


@dataclass(frozen=True)
class Map(Type):
    key: Type
    value: Type

    def __repr__(self) -> str:
        return f"Map<{self.key!r}, {self.value!r}>"


@dataclass(frozen=True)
class Option(Type):
    inner: Type

    def __repr__(self) -> str:
        return f"Option<{self.inner!r}>"


@dataclass(frozen=True)
class Param(Type):
    """The `index`th generic parameter of the enclosing record or enum."""

    index: int


class Adt(Type):
    """A record or enum, with its type arguments."""

    def __init__(self, cls: type, *args: Type) -> None:
        self.cls = cls
        self.args = args

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Adt) and (self.cls, self.args) == (other.cls, other.args)

    def __hash__(self) -> int:
        return hash((self.cls, self.args))

    def __repr__(self) -> str:
        name = info(self.cls).name
        return f"{name}<{', '.join(map(repr, self.args))}>" if self.args else name


@dataclass
class RecordInfo:
    name: str
    #: `(wardscript name, python attribute, type)`, in declaration order.
    fields: list[tuple[str, str, Type]]


@dataclass
class VariantInfo:
    name: str
    #: The `enum.Enum` member, or the variant's class for enums with payloads.
    value: _Any
    fields: list[Type]


@dataclass
class EnumInfo:
    name: str
    variants: list[VariantInfo]

    @property
    def unit_only(self) -> bool:
        return all(not v.fields for v in self.variants)


# Field lists are thunks: types may refer to classes defined later, or in modules
# that import this one.
def record(cls: type, name: str, fields: Callable[[], list[tuple[str, str, Type]]]) -> None:
    cls.__ward__ = (RecordInfo, name, fields)  # type: ignore[attr-defined]


def enum(cls: type, name: str, variants: Callable[[], list[tuple[str, _Any, list[Type]]]]) -> None:
    cls.__ward__ = (EnumInfo, name, variants)  # type: ignore[attr-defined]


def info(cls: type) -> RecordInfo | EnumInfo:
    raw = cls.__dict__.get("__ward__")
    if raw is None:
        # A variant class of an enum with payloads: its base carries the info.
        for base in cls.__mro__[1:]:
            if "__ward__" in base.__dict__:
                return info(base)
        raise TypeError(f"{cls.__name__} is not a Wardscript type")
    if isinstance(raw, tuple):
        kind, name, thunk = raw
        if kind is RecordInfo:
            raw = RecordInfo(name, list(thunk()))
        else:
            raw = EnumInfo(name, [VariantInfo(n, v, list(fs)) for n, v, fs in thunk()])
        cls.__ward__ = raw  # type: ignore[attr-defined]
    return raw


def subst(t: Type, args: tuple[Type, ...]) -> Type:
    if isinstance(t, Param):
        return args[t.index]
    if isinstance(t, List):
        return List(subst(t.elem, args))
    if isinstance(t, Map):
        return Map(subst(t.key, args), subst(t.value, args))
    if isinstance(t, Option):
        return Option(subst(t.inner, args))
    if isinstance(t, Adt):
        return Adt(t.cls, *(subst(a, args) for a in t.args))
    return t


_PRIM_SCHEMAS = {
    Int: {"type": "integer"},
    Float: {"type": "number"},
    String: {"type": "string"},
    Bool: {"type": "boolean"},
    Unit: {"type": "null"},
    Any: {},
}


def json_schema(t: Type) -> dict:
    """A JSON Schema (2020-12) for `t`. Records and enums go in `$defs`."""
    defs: dict[str, dict] = {}
    root = _schema(t, defs)
    return {**root, "$defs": defs} if defs else root


def _def_name(t: Adt) -> str:
    name = info(t.cls).name
    if not t.args:
        return name
    return "_".join([name, *(repr(a).replace("<", "_").replace(">", "").replace(", ", "_") for a in t.args)])


def _schema(t: Type, defs: dict[str, dict]) -> dict:
    if isinstance(t, Prim):
        return dict(_PRIM_SCHEMAS[t])
    if isinstance(t, List):
        return {"type": "array", "items": _schema(t.elem, defs)}
    if isinstance(t, Map):
        return {"type": "object", "additionalProperties": _schema(t.value, defs)}
    if isinstance(t, Option):
        return {"anyOf": [_schema(t.inner, defs), {"type": "null"}]}
    if isinstance(t, Adt):
        name = _def_name(t)
        if name not in defs:
            defs[name] = {}  # placeholder, so recursive types terminate
            defs[name] = _adt_schema(t, defs)
        return {"$ref": f"#/$defs/{name}"}
    raise TypeError(f"no JSON schema for {t!r}")


def _adt_schema(t: Adt, defs: dict[str, dict]) -> dict:
    i = info(t.cls)
    if isinstance(i, RecordInfo):
        return {
            "type": "object",
            "properties": {w: _schema(subst(ft, t.args), defs) for w, _, ft in i.fields},
            "required": [w for w, _, _ in i.fields],
            "additionalProperties": False,
        }
    if i.unit_only:
        return {"type": "string", "enum": [v.name for v in i.variants]}
    options: list[dict] = []
    for v in i.variants:
        if not v.fields:
            options.append({"const": v.name})
            continue
        items = [_schema(subst(ft, t.args), defs) for ft in v.fields]
        payload = {"type": "array", "prefixItems": items, "minItems": len(items), "maxItems": len(items)}
        options.append(
            {
                "type": "object",
                "properties": {v.name: payload},
                "required": [v.name],
                "additionalProperties": False,
            }
        )
    return {"oneOf": options}


def _kind(value: _Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "a boolean"
    if isinstance(value, (int, float)):
        return "a number"
    if isinstance(value, str):
        return "a string"
    if isinstance(value, list):
        return "an array"
    if isinstance(value, dict):
        return "an object"
    return type(value).__name__


def decode(t: Type, value: _Any, path: str = "$") -> _Any:
    """Converts a JSON value to the Python representation of `t`, or raises
    `DecodeError` saying where it doesn't match."""

    def fail(expected: str) -> DecodeError:
        return DecodeError(path, f"expected {expected}, found {_kind(value)}")

    if t is Any:
        return value
    if t is Int:
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise fail("an integer")
        if isinstance(value, float) and not value.is_integer():
            raise fail("an integer")
        return int(value)
    if t is Float:
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise fail("a number")
        return float(value)
    if t is String:
        if not isinstance(value, str):
            raise fail("a string")
        return value
    if t is Bool:
        if not isinstance(value, bool):
            raise fail("a boolean")
        return value
    if t is Unit:
        if value is not None:
            raise fail("null")
        return None
    if isinstance(t, List):
        if not isinstance(value, list):
            raise fail("an array")
        return [decode(t.elem, x, f"{path}[{i}]") for i, x in enumerate(value)]
    if isinstance(t, Map):
        if not isinstance(value, dict):
            raise fail("an object")
        return {
            _decode_key(t.key, k, path): decode(t.value, v, f"{path}.{k}") for k, v in value.items()
        }
    if isinstance(t, Option):
        return None if value is None else some(decode(t.inner, value, path))
    if isinstance(t, Adt):
        return _decode_adt(t, value, path, fail)
    raise TypeError(f"cannot decode {t!r}")


def _decode_key(t: Type, key: str, path: str) -> _Any:
    if t is Int:
        try:
            return int(key)
        except ValueError:
            raise DecodeError(path, f"expected integer keys, found {key!r}") from None
    return decode(t, key, path)


def _decode_adt(t: Adt, value: _Any, path: str, fail: Callable[[str], DecodeError]) -> _Any:
    i = info(t.cls)
    if isinstance(i, RecordInfo):
        if not isinstance(value, dict):
            raise fail(f"a `{i.name}` object")
        names = {w for w, _, _ in i.fields}
        for key in value:
            if key not in names:
                raise DecodeError(path, f"`{i.name}` has no field `{key}`")
        kwargs = {}
        for w, py, ft in i.fields:
            if w not in value:
                raise DecodeError(path, f"missing field `{w}` of `{i.name}`")
            kwargs[py] = decode(subst(ft, t.args), value[w], f"{path}.{w}")
        return t.cls(**kwargs)

    names = ", ".join(f"`{v.name}`" for v in i.variants)
    if isinstance(value, str):
        for v in i.variants:
            if v.name == value and not v.fields:
                return v.value if i.unit_only else v.value()
        raise DecodeError(path, f"`{value}` is not a variant of `{i.name}` without fields (variants: {names})")
    if isinstance(value, dict) and len(value) == 1 and not i.unit_only:
        [(key, payload)] = value.items()
        for v in i.variants:
            if v.name == key and v.fields:
                if not isinstance(payload, list) or len(payload) != len(v.fields):
                    raise DecodeError(f"{path}.{key}", f"expected an array of {len(v.fields)} values")
                return v.value(
                    *(decode(subst(ft, t.args), x, f"{path}.{key}[{n}]") for n, (ft, x) in enumerate(zip(v.fields, payload)))
                )
        raise DecodeError(path, f"`{key}` is not a variant of `{i.name}` with fields (variants: {names})")
    raise fail(f"a `{i.name}` variant")


def encode(value: _Any) -> _Any:
    """The JSON value for a Wardscript value; the inverse of `decode`."""
    if value is None or isinstance(value, (bool, int, float, str)):
        return value
    if isinstance(value, Some):
        return encode(unwrap(value))
    if isinstance(value, list):
        return [encode(x) for x in value]
    if isinstance(value, dict):
        return {k if isinstance(k, str) else str(encode(k)): encode(v) for k, v in value.items()}
    if isinstance(value, _enum.Enum):
        return value.name
    i = info(type(value))
    if isinstance(i, RecordInfo):
        return {w: encode(getattr(value, py)) for w, py, _ in i.fields}
    for v in i.variants:
        if type(value) is v.value:
            if not v.fields:
                return v.name
            return {v.name: [encode(getattr(value, f"_{n}")) for n in range(len(v.fields))]}
    raise TypeError(f"cannot encode {value!r}")

File: crates/ward_runtime/ts/src/schema.ts
// Runtime descriptions of Wardscript types, and what they're for: JSON schemas for
// model answers, and decoding and encoding JSON values. Records are plain objects,
// enums without fields are strings (the variant's name), and variants with fields
// are `{tag, _0, _1, ...}` objects made by `variant()`.

import { DecodeError, WardError } from "./errors.ts";
import { Some, some, unwrap } from "./values.ts";

export abstract class Type {
  abstract toString(): string;
}

export class Prim extends Type {
  readonly name: string;
  constructor(name: string) {
    super();
    this.name = name;
  }
  toString(): string {
    return this.name;
  }
}

export const Int = new Prim("Int");
export const Float = new Prim("Float");
export const String = new Prim("String");
export const Bool = new Prim("Bool");
export const Unit = new Prim("()");
/** Values of tools without a schema. */
export const Any = new Prim("dynamic");

export class ListType extends Type {
  readonly elem: Type;
  constructor(elem: Type) {
    super();
    this.elem = elem;
  }
  toString(): string {
    return `List<${this.elem}>`;
  }
}

export class MapType extends Type {
  readonly key: Type;
  readonly value: Type;
  constructor(key: Type, value: Type) {
    super();
    this.key = key;
    this.value = value;
  }
  toString(): string {
    return `Map<${this.key}, ${this.value}>`;
  }
}

export class OptionType extends Type {
  readonly inner: Type;
  constructor(inner: Type) {
    super();
    this.inner = inner;
  }
  toString(): string {
    return `Option<${this.inner}>`;
  }
}

/** The `index`th generic parameter of the enclosing record or enum. */
export class ParamType extends Type {
  readonly index: number;
  constructor(index: number) {
    super();
    this.index = index;
  }
  toString(): string {
    return `T${this.index}`;
  }
}

export interface RecordInfo {
  kind: "record";
  name: string;
  /** `[name, type]`, in declaration order. */
  fields: () => [string, Type][];
}

export interface EnumInfo {
  kind: "enum";
  name: string;
  /** `[name, payload types]`. */
  variants: () => [string, Type[]][];
}

export type AdtInfo = RecordInfo | EnumInfo;

export function record(name: string, fields: () => [string, Type][]): RecordInfo {
  return { kind: "record", name, fields };
}

export function enum_(name: string, variants: () => [string, Type[]][]): EnumInfo {
  return { kind: "enum", name, variants };
}

export class AdtType extends Type {
  readonly info: AdtInfo;
  readonly args: Type[];
  constructor(info: AdtInfo, ...args: Type[]) {
    super();
    this.info = info;
    this.args = args;
  }
  toString(): string {
    return this.args.length ? `${this.info.name}<${this.args.join(", ")}>` : this.info.name;
  }
}

/** `base where cond`: checked while decoding; `schema` holds the JSON Schema keywords
 * the condition implies. */
export class RefinedType extends Type {
  readonly base: Type;
  readonly check: (value: never) => boolean;
  readonly text: string;
  readonly schema: Record<string, unknown>;
  constructor(base: Type, check: (value: never) => boolean, text: string, schema: Record<string, unknown> = {}) {
    super();
    this.base = base;
    this.check = check;
    this.text = text;
    this.schema = schema;
  }
  toString(): string {
    return `${this.base} where ${this.text}`;
  }
}

export const List = (elem: Type): Type => new ListType(elem);
export const Adt = (info: AdtInfo, ...args: Type[]): Type => new AdtType(info, ...args);
export const Param = (index: number): Type => new ParamType(index);
export const Refined = (
  base: Type,
  check: (value: never) => boolean,
  text: string,
  schema: Record<string, unknown> = {},
): Type => new RefinedType(base, check, text, schema);
export const Map_ = (key: Type, value: Type): Type => new MapType(key, value);
export const Option = (inner: Type): Type => new OptionType(inner);

/** Marks an object as an enum variant, so encoding tells it from a record. */
export const VARIANT: unique symbol = Symbol("wardscript.variant");

export function variant<T extends { tag: string }>(value: T): T {
  Object.defineProperty(value, VARIANT, { value: true, enumerable: false });
  return value;
}

export function isVariant(value: unknown): value is { tag: string } & Record<string, unknown> {
  return value !== null && typeof value === "object" && VARIANT in (value as object);
}

export function subst(t: Type, args: Type[]): Type {
  if (t instanceof ParamType) return args[t.index] ?? Any;
  if (t instanceof ListType) return new ListType(subst(t.elem, args));
  if (t instanceof MapType) return new MapType(subst(t.key, args), subst(t.value, args));
  if (t instanceof OptionType) return new OptionType(subst(t.inner, args));
  if (t instanceof AdtType) return new AdtType(t.info, ...t.args.map((a) => subst(a, args)));
  if (t instanceof RefinedType) return new RefinedType(subst(t.base, args), t.check, t.text, t.schema);
  return t;
}

// -- JSON Schema --

type Schema = Record<string, unknown>;

export function jsonSchema(t: Type): Schema {
  const defs: Record<string, Schema> = {};
  const root = schemaOf(t, defs);
  return Object.keys(defs).length ? { ...root, $defs: defs } : root;
}

function defName(t: AdtType): string {
  if (!t.args.length) return t.info.name;
  return [t.info.name, ...t.args.map((a) => `${a}`.replace(/</g, "_").replace(/>/g, "").replace(/, /g, "_"))].join("_");
}

function schemaOf(t: Type, defs: Record<string, Schema>): Schema {
  if (t === Int) return { type: "integer" };
  if (t === Float) return { type: "number" };
  if (t === String) return { type: "string" };
  if (t === Bool) return { type: "boolean" };
  if (t === Unit) return { type: "null" };
  if (t === Any) return {};
  if (t instanceof ListType) return { type: "array", items: schemaOf(t.elem, defs) };
  if (t instanceof MapType) return { type: "object", additionalProperties: schemaOf(t.value, defs) };
  if (t instanceof OptionType) return { anyOf: [schemaOf(t.inner, defs), { type: "null" }] };
  if (t instanceof RefinedType) {
    const base = schemaOf(t.base, defs);
    const rule = `must satisfy: ${t.text}`;
    const description = typeof base.description === "string" ? `${base.description} (${rule})` : rule;
    return { ...base, ...t.schema, description };
  }
  if (t instanceof AdtType) {
    const name = defName(t);
    if (!(name in defs)) {
      defs[name] = {};
      defs[name] = adtSchema(t, defs);
    }
    return { $ref: `#/$defs/${name}` };
  }
  throw new TypeError(`no JSON schema for ${t}`);
}

function adtSchema(t: AdtType, defs: Record<string, Schema>): Schema {
  const info = t.info;
  if (info.kind === "record") {
    const fields = info.fields();
    return {
      type: "object",
      properties: Object.fromEntries(fields.map(([n, ft]) => [n, schemaOf(subst(ft, t.args), defs)])),
      required: fields.map(([n]) => n),
      additionalProperties: false,
    };
  }
  const variants = info.variants();
  if (variants.every(([, fs]) => fs.length === 0)) {
    return { type: "string", enum: variants.map(([n]) => n) };
  }
  return {
    oneOf: variants.map(([n, fs]) => {
      if (!fs.length) return { const: n };
      const items = fs.map((ft) => schemaOf(subst(ft, t.args), defs));
      return {
        type: "object",
        properties: { [n]: { type: "array", prefixItems: items, minItems: items.length, maxItems: items.length } },
        required: [n],
        additionalProperties: false,
      };
    }),
  };
}

// -- decoding --

function kind(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "boolean") return "a boolean";
  if (typeof value === "number") return "a number";
  if (typeof value === "string") return "a string";
  if (Array.isArray(value)) return "an array";
  if (typeof value === "object") return "an object";
  return typeof value;
}

/** Converts a JSON value to the JavaScript representation of `t`, or throws
 * `DecodeError` saying where it doesn't match. */
export function decode(t: Type, value: unknown, path = "$"): unknown {
  const fail = (expected: string) => new DecodeError(path, `expected ${expected}, found ${kind(value)}`);
  if (t === Any) return value;
  if (t === Int) {
    if (typeof value !== "number" || !Number.isInteger(value)) throw fail("an integer");
    return value;
  }
  if (t === Float) {
    if (typeof value !== "number") throw fail("a number");
    return value;
  }
  if (t === String) {
    if (typeof value !== "string") throw fail("a string");
    return value;
  }
  if (t === Bool) {
    if (typeof value !== "boolean") throw fail("a boolean");
    return value;
  }
  if (t === Unit) {
    if (value !== null && value !== undefined) throw fail("null");
    return null;
  }
  if (t instanceof ListType) {
    if (!Array.isArray(value)) throw fail("an array");
    return value.map((x, i) => decode(t.elem, x, `${path}[${i}]`));
  }
  if (t instanceof MapType) {
    if (value === null || typeof value !== "object" || Array.isArray(value)) throw fail("an object");
    const out = new Map<unknown, unknown>();
    for (const [k, v] of Object.entries(value as object)) {
      out.set(decodeKey(t.key, k, path), decode(t.value, v, `${path}.${k}`));
    }
    return out;
  }
  if (t instanceof OptionType) {
    return value === null || value === undefined ? null : some(decode(t.inner, value, path));
  }
  if (t instanceof RefinedType) {
    const decoded = decode(t.base, value, path);
    let ok: boolean;
    try {
      ok = (t.check as (v: unknown) => boolean)(decoded);
    } catch (e) {
      if (e instanceof WardError) throw new DecodeError(path, `checking \`${t.text}\` failed: ${e.message}`);
      throw e;
    }
    if (!ok) throw new DecodeError(path, `doesn't satisfy \`${t.text}\``);
    return decoded;
  }
  if (t instanceof AdtType) return decodeAdt(t, value, path, fail);
  throw new TypeError(`cannot decode ${t}`);
}

function decodeKey(t: Type, key: string, path: string): unknown {
  if (t === Int) {
    const n = Number(key);
    if (!Number.isInteger(n)) throw new DecodeError(path, `expected integer keys, found ${JSON.stringify(key)}`);
    return n;
  }
  return decode(t, key, path);
}

function decodeAdt(t: AdtType, value: unknown, path: string, fail: (e: string) => DecodeError): unknown {
  const info = t.info;
  if (info.kind === "record") {
    if (value === null || typeof value !== "object" || Array.isArray(value)) throw fail(`a \`${info.name}\` object`);
    const fields = info.fields();
    const names = new Set(fields.map(([n]) => n));
    for (const key of Object.keys(value as object)) {
      if (!names.has(key)) throw new DecodeError(path, `\`${info.name}\` has no field \`${key}\``);
    }
    const out: Record<string, unknown> = {};
    for (const [n, ft] of fields) {
      if (!(n in (value as object))) throw new DecodeError(path, `missing field \`${n}\` of \`${info.name}\``);
      out[n] = decode(subst(ft, t.args), (value as Record<string, unknown>)[n], `${path}.${n}`);
    }
    return out;
  }
  const variants = info.variants();
  const unitOnly = variants.every(([, fs]) => fs.length === 0);
  const names = variants.map(([n]) => `\`${n}\``).join(", ");
  if (typeof value === "string") {
    const v = variants.find(([n, fs]) => n === value && fs.length === 0);
    if (!v) throw new DecodeError(path, `\`${value}\` is not a variant of \`${info.name}\` without fields (variants: ${names})`);
    return unitOnly ? value : variant({ tag: value });
  }
  if (value !== null && typeof value === "object" && !Array.isArray(value) && !unitOnly) {
    const entries = Object.entries(value as object);
    if (entries.length === 1) {
      const [key, payload] = entries[0] as [string, unknown];
      const v = variants.find(([n, fs]) => n === key && fs.length > 0);
      if (v) {
        const fs = v[1];
        if (!Array.isArray(payload) || payload.length !== fs.length) {
          throw new DecodeError(`${path}.${key}`, `expected an array of ${fs.length} values`);
        }
        const out: Record<string, unknown> = { tag: key };
        fs.forEach((ft, n) => {
          out[`_${n}`] = decode(subst(ft, t.args), payload[n], `${path}.${key}[${n}]`);
        });
        return variant(out as { tag: string });
      }
      throw new DecodeError(path, `\`${key}\` is not a variant of \`${info.name}\` with fields (variants: ${names})`);
    }
  }
  throw fail(`a \`${info.name}\` variant`);
}

/** The JSON value for a Wardscript value; the inverse of `decode`. */
export function encode(value: unknown): unknown {
  if (value === null || value === undefined) return null;
  if (typeof value === "boolean" || typeof value === "number" || typeof value === "string") return value;
  if (value instanceof Some) return encode(unwrap(value));
  if (Array.isArray(value)) return value.map(encode);
  if (value instanceof Map) {
    return Object.fromEntries([...value].map(([k, v]) => [typeof k === "string" ? k : globalThis.String(encode(k)), encode(v)]));
  }
  if (isVariant(value)) {
    const payload = Object.keys(value).filter((k) => /^_\d+$/.test(k)).sort((a, b) => Number(a.slice(1)) - Number(b.slice(1)));
    return payload.length === 0 ? value.tag : { [value.tag]: payload.map((k) => encode(value[k])) };
  }
  if (typeof value === "object") {
    return Object.fromEntries(Object.entries(value as object).map(([k, v]) => [k, encode(v)]));
  }
  throw new TypeError(`cannot encode ${globalThis.String(value)}`);
}
